src/uu/head/src/parse.rs

```diff
@@ -0,0 +1,237 @@
+use std::convert::TryFrom;
+
+#[derive(PartialEq, Debug)]
+pub enum ParseError {
+    Syntax,
+    Overflow,
+}
+/// Parses obsolete syntax
+/// head -NUM
+/// (note:)
+pub fn parse_obsolete(src: &str) -> Option<Result<(usize, bool), ParseError>> {
+    let re = regex::Regex::new(r"^-(\d+)(?:([cbkm])+)?$").unwrap();
+    match re.captures(src) {
+        Some(cap) => {
+            let num = cap.get(1).unwrap();
+            let mut num = match num.as_str().parse::<usize>() {
+                Ok(n) => n,
+                Err(_) => {
+                    return Some(Err(ParseError::Overflow));
+                }
+            };
+            let mut bytes = false;
+            if let Some(m) = cap.get(2) {
+                let newn = match m.as_str().chars().next().unwrap_or(0 as char) {
+                    'c' => num.checked_mul(1),
+                    'b' => num.checked_mul(512),
+                    'k' => num.checked_mul(1024),
+                    'm' => num.checked_mul(1024 * 1024),
+                    _ => unreachable!(),
+                };
+                match newn {
+                    Some(n) => {
+                        num = n;
+                        bytes = true;
+                    }
+                    None => return Some(Err(ParseError::Overflow)),
+                }
+            }
+            Some(Ok((num, bytes)))
+        }
+        None => None,
+    }
+}
+/// Parses an -c or -n argument,
+/// the bool specifies whether to read from the end
+pub fn parse_num(src: &str) -> Result<(usize, bool), ParseError> {
+    // we're not using the bytefmt or byte_unit crates
+    // because they aren't fully `head` format compliant
+
+    if src.is_empty() {
+        return Err(ParseError::Syntax);
+    }
+
+    // (1) is the number
+    // (2) is b
+    // (3) is any of the other units
+    // (4) is `i`
+    // (5) is `B`
+    let re = regex::Regex::new(r"^-?(\d+)?(?:(b)|([kKmMgGtTpPeEzZyY])(?:(i)B|(B))?)?$").unwrap();
```

Comment:
I would personally prefer this not use a regex, as I think they're hard to read.  Same for `parse_obsolete()`.

Reply:
> I would personally prefer this not use a regex, as I think they're hard to read. Same for `parse_obsolete()`.

I was planning to rewrite this anyway for performance, however, I'm not sure about the readability aspect of it